### ISSUE TITLE
Definition of weighted mask in paper doesn't match with the code implementation

### ISSUE THREAD

**Issue by ghost** [moodoki/semantic_image_inpainting#2]
Hi, I am a little bit confused about the **weighted mask** in the **context loss** part, I think the definition of that in your paper and the implementation of that in your code seem to be contradictory.

So in your paper, the weighted mask or say weighting term is defined as follows.
![weighted_mask](https://user-images.githubusercontent.com/29271187/32129917-181af310-bb43-11e7-89f8-fc3c9901bdcc.PNG)
I feel it is very strange since this kind of operation will lead to an almost all-zero matrix, and that won't help to get any meaningful context loss.

Then I checked out this code, and found that in your code, the weighted mask is actually implemented as follows.
```
ker = np.ones((nsize,nsize), dtype=np.float32)
ker = ker/np.sum(ker)
wmask = mask * convolve2d(mask, ker, mode='same', boundary='symm')
```
This implementation just penalizes the importance of pixels around a hole, and won't lead to an almost all-zero matrix. It is contradictory to the definition in the paper, but I feel actually this is the right way.

I also did a very simple experiment. Assume the window size is 1, and the original mask is
```
[
    [1, 1, 1]
    [1, 0, 1]
    [1, 1, 1]
]
```
By the definition in your paper, then the weighted mask will be
```
[
    [0, 0, 0]
    [0, 0, 0]
    [0, 0, 0]
]
```
By the code implementation, then the weighted mask will be
```
[
    [1, 1, 1]
    [1, 0, 1]
    [1, 1, 1]
]
```
This is so weird, I think the definition is wrong. Can you take a look for this? Thanks in advance!

**Comment by moodoki** [comment]
Hi, I think you might have mixed up the j and i subscripts. The code implementation is as described in the paper. The intuitive interpretation of the weighted mask is to reduce the weights of pixels bordering missing regions, a neighborhood size of 1 means just considering the current pixel, therefore not changing the mask in anyway.

**Comment by ghost** [comment]
Hi, thanks for your reply, but I still cannot agree with you. According to the definition and explanation in your paper, actually the weights of pixels that are far away holes (or say missing regions) will be greatly reduced, this is contradictory to what you just said. And also, in the **prior loss** part, the paper describes how that is related to the discriminator, however in the code, you are actually using and updating generator. Your paper and code are telling different stories, as far as I am concerned. I think the code is clear and is doing the right thing, but I don't understand why the paper is ambiguous in a lot of parts, so weird.

**Comment by moodoki** [comment]
Hi, I think that there you might be confused about certain aspects in our work. Saying that our code and paper are telling different stories is a very serious allegation that should not be made lightly. (There are others who had implemented our method before we release our code publicly just based on our pre-print in arxiv.)

Firstly, we update neither the generator nor discriminator, we perform back-propagation to input to update the latent variable at each iteration. The generator and discriminator are trained separately, our method simply uses the output from these networks. (We used a DCGAN implemented by other authors.) In other words, we treat D and G as fixed, obtain a gradient w.r.t. latent variable z, and updating z to find the best latent space representation minimizing the loss as described. 

Our total loss consists of 2 components, the prior loss <a href="https://www.codecogs.com/eqnedit.php?latex=\mathcal{L}_p" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\mathcal{L}_p" title="\mathcal{L}_p" /></a> and context loss <a href="https://www.codecogs.com/eqnedit.php?latex=\mathcal{L}_c" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\mathcal{L}_c" title="\mathcal{L}_c" /></a> (see eq(2)). 

The weighted mask allows us to choose which pixels to consider when comparing the distance between the generator output and the input image and it's a hyperparameter that can be changed depending on the task at hand. This weight is applied on the generator output as described in eq(4). We call this loss the context loss.

As for the discriminator, no mask is applied, instead we applied a weighting parameter <a href="https://www.codecogs.com/eqnedit.php?latex=\lambda" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\lambda" title="\lambda" /></a> (eq(5)). This term is called the prior loss. 

In our code, you should see these 2 terms summed together as per eq(2) in our paper.

Don't hesitate to contact me or any of the other authors for clarifications if needed.

**Comment by HubYZ** [comment]
The weight of mask is not right in the master.
